smart_simulation/scale.py

```diff
@@ -61,95 +63,182 @@ def upsample(dataset, consumption_window_template: str) -> pd.DataFrame:
     return upsampled_data
 
 
-def create_weight_data(dataset: pd.DataFrame, arrival_template: str) -> pd.DataFrame:
+def calculate_arrival_windows(delivery_days_max: dict, freq: str) -> dict:
+    """
+    Calculate the arrival windows in number of time stamps for early, on-time, and late deliveries
+    Args:
+        delivery_days_max: Maximum days in early, on-time, late delivery window
+        freq: Frequency of the dataset utilized to calculate number of time stamps per day
+
+    Returns:
+        A dictionary of tuples as the range of time stamps for early, on-time, and late delivery windows
+    """
+    time_stamps_per_day = pd.Timedelta("1 day") / freq
+
+    max_early = delivery_days_max["early"]
+    early_low = max_early * time_stamps_per_day * -1
+    early_high = time_stamps_per_day * -1
+    early_range = (early_low, early_high)
+
+    max_on_time = int(delivery_days_max["on_time"] * time_stamps_per_day / 2)
+    on_time_low = max_on_time * -1
+    on_time_high = max_on_time
+    on_time_range = (on_time_low, on_time_high)
+
+    max_late = delivery_days_max["late"]
+    late_low = time_stamps_per_day
+    late_high = max_late * time_stamps_per_day
+    late_range = (late_low, late_high)
+
+    arrival_windows = {
+        "early": early_range,
+        "on_time": on_time_range,
+        "late": late_range,
+    }
+    return arrival_windows
+
+
+def deliver_product(delivery_skew_probabilities: dict, arrival_windows: dict) -> int:
+    """
+    Calculate and return the delta, in number of timestamps, of delivery from a true 'scale weight == zero' timestamp
+    Args:
+        delivery_skew_probabilities: A dictionary of of probabilities for early, on-time, and late delivery
+        arrival_windows: A dictionary of acceptable range in timestamps for early, on-time, and late delivery
+
+    Returns: A int representing the delta of timestamps away from true 'scale weight == zero' that a delivery arrives
+
+    """
+    arrival_time_categories = [*delivery_skew_probabilities]
+    arrival_time_probabilities = tuple(delivery_skew_probabilities.values())
+    # random.choices() returns a list of 'k' number selections, just need string value
+    arrival_time_category = random.choices(
+        population=arrival_time_categories, weights=arrival_time_probabilities, k=1,
```

Comment:
the `weights` parameter can take a list or a tuple.

Reply:
Changed to a list for consistency with `arrival_time_categories`